### PR TITLE
Feat/expiration

### PR DESCRIPTION
**Pull request by wsy19961129**
Thins to be confirmed (comments on whether these settings are resonable):
    1. Some basic data types are changed, i.e., createdAt, expiredAt, locked in IUserWithoutToken;
    2. Add a new page manageBatchUsers such that the admin could lock/unlock users within specific projects(datasets later);

If these are confirmed, the next steps are:
    1. Complete these functions;
    2. Update all old test files and write new test files for batch operation.

### CODE REVIEW

File: packages/itmat-ui-react/src/components/users/index.tsx [hunk 1]
```diff
@@ -26,6 +27,14 @@ export const UserPage: React.FunctionComponent = (props) => {
                             </div>
                         </>
                     } />
+                    <Route path="/users/manageBatchUsers" render={() =>
+                        <>
+                            <div className="page_ariane">CREATE NEW USER</div>
```
Comment (fguitton) [inline packages/itmat-ui-react/src/components/users/index.tsx:32]: Why does this have the same title as the other category ?

File: packages/itmat-ui-react/src/components/users/manageBatchUsers.tsx [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import { Query, Mutation } from 'react-apollo';
+import { NavLink } from 'react-router-dom';
+import { WHO_AM_I } from 'itmat-commons/dist/graphql/user';
+import css from './userList.module.css';
+import { useQuery } from 'react-apollo';
+import { GET_PROJECT } from 'itmat-commons/dist/graphql/projects';
+import { LoadingBalls } from '../reusable/icons/loadingBalls';
+import { EDIT_USER } from 'itmat-commons/dist/graphql';
+
+export const ManageBatchUsers: React.FunctionComponent = (props) => {
```
Comment (fguitton) [inline packages/itmat-ui-react/src/components/users/manageBatchUsers.tsx:12]: I am not sure why we need a new page. I'm not clear what it's supposed to do considering there already exists a view for managing a users.

File: packages/itmat-interface/src/graphql/core/userCore.ts [hunk 3]
```diff
@@ -17,8 +17,8 @@ export class UserCore {
         return user;
     }
 
-    public async createUser(requester: string, user: { password: string, username: string, organisation: string, type: userTypes, description: string, realName: string, email: string, emailNotificationsActivated: boolean }): Promise<IUserWithoutToken> {
-        const { password, organisation, username, type, description, realName, email, emailNotificationsActivated } = user;
+    public async createUser(requester: string, user: { password: string, username: string, organisation: string, type: userTypes, description: string, realName: string, email: string, emailNotificationsActivated: boolean, createdAt: number, expiredAt: number, locked: boolean }): Promise<IUserWithoutToken> {
```
Comment (fguitton) [inline packages/itmat-interface/src/graphql/core/userCore.ts:20]: It in unclear why we would need to pass in `createdAt` in the function that will create the user. Isn't that something the function should derive on its own ?

File: packages/itmat-commons/src/models/user.ts [hunk 4]
```diff
@@ -14,6 +14,9 @@ export interface IUserWithoutToken {
     emailNotificationsActivated: boolean;
     deleted: number | null;
     createdBy: string;
+    createdAt: number;
+    expiredAt: number;
+    locked: boolean
```
Comment (fguitton) [inline packages/itmat-commons/src/models/user.ts:19]: I don't understand how this flag is necessary. i.e. `if currentTime > expiredAt -> locked`

Comment (sou-chon) [reply packages/itmat-commons/src/models/user.ts:19]: exactly

Comment (sou-chon) [reply packages/itmat-commons/src/models/user.ts:19]: does "locked" mean something else? But would be good to have a function where admin can "disable" a user at will, like if a user is transferred to another project for a while (but might come back). would be good to "freeze" their account but can revive it later with all the permission and user info intact. 

But this can be implemented by setting expireAt to 0 when admin "freezes" the account. so might be better to get rid of "locked" flag anyway

File: packages/itmat-ui-react/src/components/users/userDetails.tsx [hunk 5]
```diff
@@ -81,6 +81,9 @@ export const EditUserForm: React.FunctionComponent<{ user: (IUserWithoutToken &
                     <label>Description:  <input type='text' value={inputs.description} onChange={e => { setInputs({ ...inputs, description: e.target.value }) }} /></label> <br /><br />
                     <label>Organisation: <input type='text' value={inputs.organisation} onChange={e => setInputs({ ...inputs, organisation: e.target.value })} /> </label><br /><br />
                     <label>Created by (readonly): <input type='text' readOnly value={inputs.createdBy} /> </label><br /><br />
+                    <label>Created at (readOnly): <input type='text' readOnly value={(new Date(inputs.createdAt)).toLocaleString()} /></label><br /><br />
+                    <label>Expired at: <input type='datetime-local' value={(new Date(inputs.expiredAt)).toISOString().substring(0, 16) } onChange={e => { setInputs({ ...inputs, expiredAt: (new Date(e.target.value)).getTime() }) }} /></label><br /><br />
```
Comment (sou-chon) [inline packages/itmat-ui-react/src/components/users/userDetails.tsx:85]: just a thought - is it better to have:
1. uniformly a expiry window "say 2 months", after which admin has to revive; or
2. let admin select a date (e.g. 20-7-2020) explicitly; or
3. let admin select a relative date (e.g after 20 days) explicitly

open to suggstion

**Comment by sou-chon** [comment]
Also, would be nice to have a mechanism for user to request reactivation - which will be approved by admin. because it seems very time consuming for admin to go to "edit user" user by user. 

also we don't want the user emailing the admin on their own (like they do with me now with transmart)

Would be nice if the user, after trying to login with expired credentials, gets redirected to a page where they can click "request reactivation" (or even just a button that appears after they fail to login), which then either send the admin an email / have admin notification (not implemented yet)

But this is not urgent now

File: packages/itmat-ui-react/src/components/users/manageBatchUsers.tsx (hunk 2 above)
Comment (sou-chon) [reply packages/itmat-ui-react/src/components/users/manageBatchUsers.tsx:12]: seconded

File: packages/itmat-commons/src/models/user.ts (hunk 4 above)
Comment (wsy19961129) [reply packages/itmat-commons/src/models/user.ts:19]: The reason to set both 'expiredAt' and locked is, based on my understanding, e.g., a project may last 2 months, thus the 'expiredAt' would be 2 months later. But if the admin want to review/modify the projects during this time,  like system maintenance, it would be good to lock all users temporarily such that they would not change anything.

File: packages/itmat-ui-react/src/components/users/index.tsx (hunk 1 above)
Comment (wsy19961129) [reply packages/itmat-ui-react/src/components/users/index.tsx:32]: Sorry it is a mistake, would correct it.

File: packages/itmat-ui-react/src/components/users/manageBatchUsers.tsx (hunk 2 above)
Comment (wsy19961129) [reply packages/itmat-ui-react/src/components/users/manageBatchUsers.tsx:12]: As I found that each time the admin could only operate on a single user. Like what I said below about the difference between 'expiredAt' and 'locked'. I think it would be more convenient for the admin to lock a group of users under some conditions, e.g., if he/she want to maintain a project, then the users involved in this project should be all locked temporarily.

File: packages/itmat-ui-react/src/components/users/userDetails.tsx (hunk 5 above)
Comment (wsy19961129) [reply packages/itmat-ui-react/src/components/users/userDetails.tsx:85]: Sure, would complete that, thanks!!!

File: packages/itmat-interface/src/graphql/core/userCore.ts (hunk 3 above)
Comment (wsy19961129) [reply packages/itmat-interface/src/graphql/core/userCore.ts:20]: Sure, would correct that.

File: packages/itmat-ui-react/src/components/users/userDetails.tsx (hunk 5 above)
Comment (sou-chon) [reply packages/itmat-ui-react/src/components/users/userDetails.tsx:85]: no i mean, we could pick one of these options

Comment (sou-chon) [reply packages/itmat-ui-react/src/components/users/userDetails.tsx:85]: @fguitton thoughts

File: packages/itmat-commons/src/models/user.ts (hunk 4 above)
Comment (sou-chon) [reply packages/itmat-commons/src/models/user.ts:19]: if you are looking to lock all users, the lock should be at the app level right? but i dont think we need that right now. 

I like that the expiry date to be linked to a project though; but then we have to implement a expiry date for the project as well. also a user might be linked to multiple project

File: packages/itmat-ui-react/src/components/users/manageBatchUsers.tsx (hunk 2 above)
Comment (sou-chon) [reply packages/itmat-ui-react/src/components/users/manageBatchUsers.tsx:12]: true, but in that case i would think we should make a user group entity like in *NIX systems to allow managing permissions in group, instead of linking a "group" to a project by filtering projectID.

We should think if the 'user group' concept is really needed though. The only thing you need to edit users in batch is permission. You will not really edit other fields in batch. 
we already have roles attached to each project and each studies. Each role contains users.
Wouldn't it be  easier and more straight forward to implement a "freeze" role function? so if a role is frozen, the user can't perform actions that are permitted under such role. but then they can still perform other stuff if they have other roles.

**Comment by wsy19961129** [comment]
I have removed all extra flags and pages that we thought to be uncessary. Thus the admin can set compulsory expiration time for each user, sorry to make things complex.  The exact format is to be discussed as [Chon suggested](https://github.com/dsi-icl/itmat-broker/pull/502#discussion_r431358521).

I am planning to implement the following functions according to our discussion:
1. Reactivate a locked account. Probably need to add an interface to save all such requests, and add a block/a new page to confirm them for admins;
2. Besides the account expiration date, the project will also have a expiration date such that when the expiration date of a project passes, the project cannot be accessed by STANDARD users; 

Any comments will be appreciated!

**Comment by wsy19961129** [comment]
I have removed all extra flags and pages that we thought to be uncessary. Thus the admin can set compulsory expiration time for each user, sorry to make things complex. The exact format is to be discussed as Chon suggested.

I am planning to implement the following functions according to our discussion:

Reactivate a locked account. Probably need to add an interface to save all such requests, and add a block/a new page to confirm them for admins;
Besides the account expiration date, the project will also have a expiration date such that when the expiration date of a project passes, the project cannot be accessed by STANDARD users;

Any comments will be appreciated! @sou-chon @fguitton

File: packages/itmat-interface/src/graphql/resolvers/userResolvers.ts [hunk 6]
```diff
@@ -365,7 +369,8 @@ export const userResolvers = {
                 emailNotificationsActivated,
                 password,
                 description,
-                organisation
+                organisation,
+                expiredAt,
```
Comment (fguitton) [inline packages/itmat-interface/src/graphql/resolvers/userResolvers.ts:373]: There is no need for the trailing comma

File: packages/itmat-interface/test/serverTests/study.test.ts [hunk 7]
```diff
@@ -1038,8 +1050,10 @@ describe('STUDY API', () => {
                         id: `user_access_obj_user_id_${createdUserAuthorisedStudy.id}`,
                         projects: [],
                         studies: []
-                    }
+                    },
                 });
+                expect((res.body.data.createUser.expiredAt - 
+                    res.body.data.createUser.createdAt) / 1000).toBeCloseTo(86400.0, 1);
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/study.test.ts:1056]: Is there a reason not to keep the comparison in milliseconds ?

File: packages/itmat-interface/test/serverTests/study.test.ts [hunk 8]
```diff
@@ -1230,6 +1246,8 @@ describe('STUDY API', () => {
                         }]
                     }
                 });
+                expect((res.body.data.whoAmI.expiredAt - 
+                    res.body.data.whoAmI.createdAt) / 1000).toBeCloseTo(400000000.0, 1);
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/study.test.ts:1250]: Is there a reason not to keep the comparison in milliseconds ?

File: packages/itmat-utils/src/databaseSetup/seed/users.js [hunk 9]
```diff
@@ -12,6 +12,8 @@ const users = [{
     "emailNotificationsActivated" : false, 
     "deleted" : null, 
     "id" : "replaced_at_runtime2", 
+    "createdAt": 1591134065000,
+    "expiredAt": 1991134065000,
```
Comment (fguitton) [inline packages/itmat-utils/src/databaseSetup/seed/users.js:16]: There is no need for the trailing comma

File: packages/itmat-utils/src/databaseSetup/seed/users.js [hunk 10]
```diff
@@ -27,6 +29,8 @@ const users = [{
     "organisation":  "DSI",
     "deleted" : null, 
     "id" : "replaced_at_runtime1",
+    "createdAt": 1591134065000,
+    "expiredAt": 1991134065000,
```
Comment (fguitton) [inline packages/itmat-utils/src/databaseSetup/seed/users.js:33]: There is no need for the trailing comma

File: yarn.lock [hunk 11]
```diff
@@ -962,13 +962,16 @@
   version "2.16.2"
   resolved "https://registry.yarnpkg.com/@octokit/types/-/types-2.16.2.tgz#4c5f8da3c6fecf3da1811aef678fda03edac35d2"
   integrity sha512-O75k56TYvJ8WpAakWwYRN8Bgu60KrmX0z1KqFp1kNiFNkgW+JW+9EBKZ+S33PU6SLvbihqd+3drvPxKK68Ee8Q==
+<<<<<<< HEAD
```
Comment (fguitton) [inline yarn.lock:965]: This is clearly a merging error. This needs to be corrected. Regenerating the whole file is probably preferable.

Comment (fguitton) [reply yarn.lock:965]: This comment applies for other occurrences in this file.

File: packages/itmat-interface/test/serverTests/users.test.ts [hunk 12]
```diff
@@ -991,9 +1120,11 @@ describe('USERS API', () => {
                         id: `user_access_obj_user_id_${createdUser.id}`,
                         projects: [],
                         studies: []
-                    }
+                    },
                 }
-            )
+            );
+            expect((res.body.data.createUser.expiredAt - 
+                res.body.data.createUser.createdAt) / 1000).toBeCloseTo(86400.0, 1);
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/users.test.ts:1127]: Is there a reason not to keep the comparison in milliseconds ?

File: packages/itmat-interface/test/serverTests/users.test.ts [hunk 13]
```diff
@@ -1050,7 +1183,7 @@ describe('USERS API', () => {
                     type: userTypes.STANDARD
                 }
             });
-
+
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/users.test.ts:1186]: These whitespaces should not be there.

File: packages/itmat-interface/test/serverTests/users.test.ts [hunk 14]
```diff
@@ -991,9 +1120,11 @@ describe('USERS API', () => {
                         id: `user_access_obj_user_id_${createdUser.id}`,
                         projects: [],
                         studies: []
-                    }
+                    },
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/users.test.ts:1123]: There is no need for the trailing comma

File: packages/itmat-interface/test/serverTests/users.test.ts [hunk 15]
```diff
@@ -780,7 +904,8 @@ describe('USERS API', () => {
                     createdBy: 'admin', 
                     organisation: 'DSI',
                     id: userId
-                }
+                },
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/users.test.ts:907]: There is no need for the trailing comma

File: packages/itmat-interface/test/serverTests/users.test.ts [hunk 16]
```diff
@@ -965,7 +1094,7 @@ describe('USERS API', () => {
                     organisation: 'DSI-ICL',
                     emailNotificationsActivated: false,
                     email: 'fake@email.io',
-                    type: userTypes.STANDARD
+                    type: userTypes.STANDARD,
```
Comment (fguitton) [inline packages/itmat-interface/test/serverTests/users.test.ts:1097]: There is no need for the trailing comma

File: packages/itmat-commons/src/graphql/appUsers.ts [hunk 17]
```diff
@@ -12,6 +12,8 @@ export const GET_USERS = gql`
             email @include (if: $fetchDetailsAdminOnly)
             createdBy
             organisation
+            createdAt @include (if: $fetchDetailsAdminOnly)
+            expiredAt @include (if: $fetchDetailsAdminOnly)
```
Comment (fguitton) [inline packages/itmat-commons/src/graphql/appUsers.ts:16]: @sou-chon Do we think this makes sense from a grammatical perspective ?

File: packages/itmat-interface/test/serverTests/study.test.ts (hunk 7 above)
Comment (wsy19961129) [reply packages/itmat-interface/test/serverTests/study.test.ts:1056]: Cause I found when testing on MacOS, the error canbe 1ms sometimes, which is on the digit instead of the decimal part.

File: packages/itmat-ui-react/src/components/users/userDetails.tsx [hunk 18]
```diff
@@ -73,6 +74,34 @@ export const EditUserForm: React.FunctionComponent<{ user: (IUserWithoutToken &
...
     if (whoamierror) { return <p>ERROR: please try again.</p>; }
 
+    const showTimeFunc = {
+        showDate: function(timestamps: number) {
+            return new Date(timestamps).toISOString().substring(0, 10);
+        },
+        showTime: function(timestamps: number) {
+            return new Date(timestamps).toISOString().substring(11, 19);
+        }
+    }
```
Comment (sou-chon) [inline packages/itmat-ui-react/src/components/users/userDetails.tsx:84]: Not entirely sure but - i think from a optimisation perspective these pure functions should declared outside the function scope. or else every time this component re-renders these functions need to be recompiled and assigned. 

Not sure if the JIT compiler is smart enough to cache the result? i might be wrong

File: packages/itmat-interface/src/graphql/core/userCore.ts [hunk 19]
```diff
@@ -18,8 +18,10 @@ export class UserCore {
     }
 
     public async createUser(requester: string, user: { password: string, otpSecret: string, username: string, organisation: string, type: userTypes, description: string, realName: string, email: string, emailNotificationsActivated: boolean }): Promise<IUserWithoutToken> {
-        const { password, otpSecret, organisation, username, type, description, realName, email, emailNotificationsActivated } = user;
+        const { password, otpSecret, organisation, username, type, description, realName, email, emailNotificationsActivated  } = user;
         const hashedPassword: string = await bcrypt.hash(password, config.bcrypt.saltround);
+        const createdAt = new Date().getTime();
+        const expiredAt = new Date().getTime() + 86400 * 1000;
```
[the remark this answers is not in the document]
Comment (wsy19961129) [reply packages/itmat-interface/src/graphql/core/userCore.ts:23]: Sure

File: packages/itmat-interface/test/serverTests/study.test.ts [hunk 20]
```diff
@@ -927,7 +937,7 @@ describe('STUDY API', () => {
                 expect(res.status).toBe(200);
                 expect(res.body.errors).toBeUndefined();
                 createdUserAuthorised = await mongoClient.collection(config.database.collections.users_collection).findOne({ username });
-                expect(res.body.data.createUser).toEqual({
+                expect(res.body.data.createUser).toMatchObject({
```
[the remark this answers is not in the document]
Comment (wsy19961129) [reply packages/itmat-interface/test/serverTests/study.test.ts:940]: Because the values of $createdAt and $expiredAt are set to the instant time, i,e., dynamically. Thus I cannot know the exactly value of it.

File: packages/itmat-interface/test/serverTests/study.test.ts [hunk 21]
```diff
@@ -953,8 +957,8 @@ describe('STUDY API', () => {
                         studies: []
                     }
                 });
-                expect((res.body.data.createUser.expiredAt - 
-                    res.body.data.createUser.createdAt) / 1000).toBeCloseTo(86400.0, 1);
+                expect(Math.abs(res.body.data.createUser.expiredAt - 
+                    res.body.data.createUser.createdAt - oneDayDiff) < minTimeError).toBeTruthy();
```
Comment (sou-chon) [inline packages/itmat-interface/test/serverTests/study.test.ts:961]: I mean like
```javascript
expect(parseInt(expiredAt) - parseInt(createdAt)).toBe(oneDayDiff)
```

does this work?

Comment (wsy19961129) [reply packages/itmat-interface/test/serverTests/study.test.ts]: Thanks, Chon. It works now.

File: packages/itmat-interface/test/serverTests/study.test.ts [hunk 22]
```diff
@@ -29,6 +29,7 @@ import setupDatabase from 'itmat-utils/src/databaseSetup/collectionsAndIndexes';
 import config from '../../config/config.sample.json';
 import { v4 as uuid } from 'uuid';
 import { text } from 'body-parser';
+import { parse } from 'querystring';
```
Comment (sou-chon) [inline packages/itmat-interface/test/serverTests/study.test.ts:32]: why is this here?

File: packages/itmat-interface/test/serverTests/study.test.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (sou-chon) [reply packages/itmat-interface/test/serverTests/study.test.ts:943]: @wsy19961129 
mock the Date.now() function like this:
[https://github.com/facebook/jest/issues/2234](https://github.com/facebook/jest/issues/2234)

toMatchObject() is less strict than toEqual()
we do not want to compromise on that. it can miss future errors

File: packages/itmat-interface/src/graphql/core/userCore.ts [hunk 23]
```diff
@@ -18,8 +18,10 @@ export class UserCore {
     }
 
     public async createUser(requester: string, user: { password: string, otpSecret: string, username: string, organisation: string, type: userTypes, description: string, realName: string, email: string, emailNotificationsActivated: boolean }): Promise<IUserWithoutToken> {
-        const { password, otpSecret, organisation, username, type, description, realName, email, emailNotificationsActivated } = user;
+        const { password, otpSecret, organisation, username, type, description, realName, email, emailNotificationsActivated  } = user;
```
[the remark this answers is not in the document]
Comment (sou-chon) [reply packages/itmat-interface/src/graphql/core/userCore.ts]: There's still a space here
